src/uu/cksum/src/cksum.rs

```diff
@@ -5,134 +5,188 @@
 //  For the full copyright and license information, please view the LICENSE
 // file that was distributed with this source code.
 
-// spell-checker:ignore (ToDO) fname
+// spell-checker:ignore (ToDO) fname, algo
 use clap::{crate_version, Arg, Command};
+use hex::encode;
+use std::ffi::OsStr;
 use std::fs::File;
 use std::io::{self, stdin, BufReader, Read};
+use std::iter;
 use std::path::Path;
-use uucore::display::Quotable;
-use uucore::error::{FromIo, UResult};
-use uucore::{format_usage, show};
-
-// NOTE: CRC_TABLE_LEN *must* be <= 256 as we cast 0..CRC_TABLE_LEN to u8
-const CRC_TABLE_LEN: usize = 256;
-const CRC_TABLE: [u32; CRC_TABLE_LEN] = generate_crc_table();
+use uucore::{
+    error::{FromIo, UResult},
+    format_usage,
+    sum::{
+        div_ceil, Blake2b, Digest, DigestWriter, Md5, Sha1, Sha224, Sha256, Sha384, Sha512, Sm3,
+        BSD, CRC, SYSV,
+    },
+};
 
 const USAGE: &str = "{} [OPTIONS] [FILE]...";
 const ABOUT: &str = "Print CRC and size for each file";
 
-const fn generate_crc_table() -> [u32; CRC_TABLE_LEN] {
-    let mut table = [0; CRC_TABLE_LEN];
-
-    let mut i = 0;
-    while i < CRC_TABLE_LEN {
-        table[i] = crc_entry(i as u8);
-
-        i += 1;
-    }
-
-    table
-}
-
-const fn crc_entry(input: u8) -> u32 {
-    let mut crc = (input as u32) << 24;
-
-    let mut i = 0;
-    while i < 8 {
-        let if_condition = crc & 0x8000_0000;
-        let if_body = (crc << 1) ^ 0x04c1_1db7;
-        let else_body = crc << 1;
-
-        // NOTE: i feel like this is easier to understand than emulating an if statement in bitwise
-        //       ops
-        let condition_table = [else_body, if_body];
-
-        crc = condition_table[(if_condition != 0) as usize];
-        i += 1;
+fn detect_algo(program: &str) -> (&'static str, Box<dyn Digest + 'static>, usize) {
+    match program {
+        "sysv" => ("SYSV", Box::new(SYSV::new()) as Box<dyn Digest>, 512),
+        "bsd" => ("BSD", Box::new(BSD::new()) as Box<dyn Digest>, 1024),
+        "crc" => ("CRC", Box::new(CRC::new()) as Box<dyn Digest>, 256),
+        "md5" => ("MD5", Box::new(Md5::new()) as Box<dyn Digest>, 128),
+        "sha1" => ("SHA1", Box::new(Sha1::new()) as Box<dyn Digest>, 160),
+        "sha224" => ("SHA224", Box::new(Sha224::new()) as Box<dyn Digest>, 224),
+        "sha256" => ("SHA256", Box::new(Sha256::new()) as Box<dyn Digest>, 256),
+        "sha384" => ("SHA384", Box::new(Sha384::new()) as Box<dyn Digest>, 384),
+        "sha512" => ("SHA512", Box::new(Sha512::new()) as Box<dyn Digest>, 512),
+        "blake2b" => ("BLAKE2", Box::new(Blake2b::new()) as Box<dyn Digest>, 512),
+        "sm3" => ("SM3", Box::new(Sm3::new()) as Box<dyn Digest>, 512),
+        _ => panic!("unknown algorithm"),
```

Comment:
One more thing. This is probably not a great error to show to the user. We can restrict the values clap accepts with `Arg::possible_values`, so that this panic will never be reached.

Reply:
but I need a default option in match.

Reply:
Yeah the `_` case will need to be there still, but we can change it to `unreachable!("unknown algorithm: clap should have prevented this case")`

Reply:
done